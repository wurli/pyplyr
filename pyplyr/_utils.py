import numpy as np

def _as_id(x: np.ndarray, axis=0):
    _, id = np.unique(x, return_inverse=True, axis=axis)
    return id


def _reorder(l, items, after = None):

    after = -1 if after is None else after % len(l)

    start = []
    middle = []
    end = []
    for i, x in enumerate(l):
        if x in items:
            middle.append(x)
        elif i <= after:
            start.append(x)
        else:
            end.append(x)
    return start + middle + end


def _make_col(dfs, x):
    if not callable(x):
        return _to_vector(x)

    input_cols = list(x.__code__.co_varnames)
    existing_cols = {c for df in dfs for c in df.keys()}

    bad_args = [col for col in input_cols if col not in existing_cols]
    if len(bad_args) > 0:
        bad_args = ", ".join(bad_args)
        raise ValueError(f"Bad input: non-existent column(s) `{bad_args}`")
    
    def get_col(c):
        for df in dfs:
            if c in df.keys():
                return df[c]
    
    out = x(**{c: get_col(c) for c in input_cols})
    return _to_vector(out)


def _to_vector(x):
    return np.asarray(x).reshape(-1)
    
    
def _rm_none(l):
    return [x for x in l if x is not None]