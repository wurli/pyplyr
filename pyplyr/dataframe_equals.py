import numpy as np

def dataframe_equals(x, y):
    x_cols = x.colnames()
    y_cols = y.colnames()
    
    if x_cols != y_cols:
        return False
    
    for col in x_cols:
        if not np.array_equal(x[col], y[col]):
            return False
    
    return True