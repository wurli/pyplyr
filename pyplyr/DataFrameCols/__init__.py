from .._utils import _reorder
import copy    
import numpy as np

class DataFrameCols(dict):
    def __init__(self, **kwargs):
        super().__init__(**kwargs)


    def __getitem__(self, pos):
        if pos is None:
            return self

        pos = self.as_colname(pos)
        
        if isinstance(pos, list) or isinstance(pos, np.ndarray):
            return DataFrameCols(**{p: self[p] for p in pos if pos})
        
        return super().__getitem__(pos)


    def reorder(self, cols, after = None):

        # TODO: check all keys exist
        # TODO: check after is scalar
        after = self.as_colnum(after)

        new_cols = _reorder(self.colnames(), cols, after)

        return self._new(**{col: self[col] for col in new_cols})


    def colnames(self):
        return list(self.keys())
    
    
    def ncol(self):
        return len(self.colnames())
    

    def as_colname(self, i):
        if i is None or isinstance(i, str):
            return i
        if isinstance(i, int):
            return self.colnames()[i]
        if isinstance(i, list) or isinstance(i, np.ndarray):
            i = [ii for ii in i if ii is not None]
            return [self.as_colname(ii) for ii in i]
        raise KeyError(f"Cannot get index of type {type(i)}")
    

    def as_colnum(self, i):
        if i is None or isinstance(i, int):
            return i
        if isinstance(i, str):
            return self.colnames().index(i)
        if isinstance(i, list):
            return [self.as_colname(ii) for ii in i]
        raise KeyError(f"Cannot get get index of type {type(i)}")
    
    
    def copy(self):
        return self.__class__(**copy.copy(dict(self)))
    
        
    def _new(self, **kwargs):
        return self.__class__(**kwargs)