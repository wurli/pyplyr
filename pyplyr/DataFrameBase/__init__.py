import numpy as np
from ..DataFrameCols import DataFrameCols

class DataFrameBase(DataFrameCols):
    def __init__(self, **kwargs):
        arrays = np.broadcast_arrays(*list(kwargs.values()))
        self = super().__init__(**{
            key: arr for key, arr in zip(kwargs.keys(), arrays)
        })
    
    from .__repr__    import __repr__
    from .__getitem__ import __getitem__, _subset_rows


    # TODO: implement key: list
    def __setitem__(self, key, value):
        key = self.as_colname(key)
        old_cols = self.colnames()
        col_is_new = key not in old_cols
        DataFrameCols.__setitem__(self, key, self._make_col(value))

        if col_is_new:
            self = self.reorder(key, after=-1)
        else:
            self = self.reorder(old_cols)
    

    def _check_exists(self, *args):
        non_existant = set(args) - set(self.colnames())
        if len(non_existant) > 0:
            non_existant = ", ".join(non_existant)
            raise KeyError(f"Non-existant column(s): check {non_existant}")
    

    def _make_col(self, x):
        _, x = np.broadcast_arrays(self[0], x)
        return x


    def nrow(self):
        return len(self[0])

    
