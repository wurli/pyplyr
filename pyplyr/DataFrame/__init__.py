from ..DataFrameBase import DataFrameBase


class DataFrame(DataFrameBase):
    def __init__(self, **args):
        super().__init__(**args)

    from .mutate import mutate

    