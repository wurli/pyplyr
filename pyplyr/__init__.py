from .DataFrame import DataFrame
from .DataFrameBase import DataFrameBase
from .DataFrameCols import DataFrameCols

__all__ = [
    "DataFrame",
    "DataFrameBase",
    "DataFrameCols"
]