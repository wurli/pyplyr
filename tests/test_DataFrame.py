import pyplyr as pyp
import numpy as np
import numpy.testing as npt

#Â Snapshots using syrupy
# Use pytest --snapshot-update to update snapshots

def test_df_creation():
    df = pyp.DataFrame(
        x=['foo', 'foo', 'bar'],
        y=[11.22, 2.3, 3.4],
        z=[3, 4, 5]
    )
    
    npt.assert_array_equal(df['x'], np.array(['foo', 'foo', 'bar']))
    npt.assert_array_equal(df['y'], np.array([11.22, 2.3, 3.4]))
    npt.assert_array_equal(df['z'], np.array([3, 4, 5]))

def test_df_printing(capsys, snapshot):
    df = pyp.DataFrame(
        x=['foo', 'foo', 'bar'],
        y=[11.22, 2.3, 3.4],
        z=[3, 4, 5]
    ) 
    
    print(df)
    
    captured = capsys.readouterr().out
    assert captured == snapshot
    

def test_empty_df_printing(capsys, snapshot):
    df_nocols = pyp.DataFrame() 
    df_norows = pyp.DataFrame(x=[], y=[])
    
    print(df_nocols)
    print(df_norows)
    
    captured = capsys.readouterr().out
    assert captured == snapshot