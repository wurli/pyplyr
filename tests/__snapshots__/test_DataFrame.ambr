# serializer version: 1
# name: test_df_printing
  '''
  # A DataFrame: 3 x 3
    x               y       z
    <str96> <float64> <int64>
  0 foo         11.22       3
  1 foo           2.3       4
  2 bar           3.4       5
  
  
  '''
# ---
# name: test_empty_df_printing
  '''
  # A DataFrame: 0 x 0
  
  # A DataFrame: 0 x 2
            x         y
    <float64> <float64>
  
  
  '''
# ---
